libraries/Microsoft.Bot.Builder.LanguageGeneration/StaticChecker.cs

```diff
@@ -3,501 +3,400 @@
+            this.templates = templates;
+            baseExpressionEngine = expressionEngine ?? new ExpressionEngine();
         }
 
-        public List<Diagnostic> CheckFiles(IEnumerable<string> filePaths, ImportResolverDelegate importResolver = null)
+        // Create a property because we want this to be lazy loaded
+        private IExpressionParser ExpressionParser
         {
-            var result = new List<Diagnostic>();
-            var templates = new List<LGTemplate>();
-            var isParseSuccess = true;
-            try
+            get
             {
-                var totalLGResources = new List<LGResource>();
-                foreach (var filePath in filePaths)
+                if (_expressionParser == null)
                 {
-                    // do not use ??=, it will cause issue in the C# < 8.0
-                    importResolver = importResolver ?? ImportResolver.FileResolver;
-
-                    var fullPath = Path.GetFullPath(ImportResolver.NormalizePath(filePath));
-                    var rootResource = LGParser.Parse(File.ReadAllText(fullPath), fullPath);
-                    var resources = rootResource.DiscoverDependencies(importResolver);
-                    totalLGResources.AddRange(resources);
+                    // create an evaluator to leverage it's customized function look up for checking
+                    var evaluator = new Evaluator(templates, baseExpressionEngine);
+                    _expressionParser = evaluator.ExpressionEngine;
                 }
 
-                var deduplicatedLGResources = totalLGResources.GroupBy(x => x.Id).Select(x => x.First()).ToList();
-                templates = deduplicatedLGResources.SelectMany(x => x.Templates).ToList();
+                return _expressionParser;
             }
-            catch (LGException ex)
+        }
+
+        /// <summary>
+        /// Return error messaages list.
+        /// </summary>
+        /// <returns>report result.</returns>
+        public List<Diagnostic> Check()
+        {
+            var result = new List<Diagnostic>();
+
+            // check dup first
+            var duplicatedTemplates = templates
+                                      .GroupBy(t => t.Name)
+                                      .Where(g => g.Count() > 1)
+                                      .Where(u => !string.IsNullOrEmpty(u.Key))
+                                      .ToList();
+
+            if (duplicatedTemplates.Count > 0)
             {
-                result.AddRange(ex.Diagnostics);
-                isParseSuccess = false;
+                duplicatedTemplates.ForEach(g =>
+                {
+                    var name = g.Key;
+                    var sources = string.Join(":", g.Select(x => x.Source));
+
+                    var msg = $"Duplicated definitions found for template: {name} in {sources}";
+                    result.Add(BuildLGDiagnostic(msg));
```

Comment:
i don't think the BuildLGDianostic here include source?

Reply:
Perhaps source should be a required field, to make things  easier?

Reply:
Never mind, this source is diagnostics should not be used as location definition.